### PR TITLE
[PIP 81]: Part-2 Add range set wrapper to support record dirty ledgers

### PR DESCRIPTION
**Pull request by 315157973**
### Motivation
Pulsar persists the individual acknowledgments into Ledger for each subscription.
acknowledgment hole uses guava range to record.
Now a single entry is used to record these holes, but a single entry can not stores more than 5MB data.
Therefore, this PIP stores these holes in multiple entries separately. And use LRU to ensure low memory usage.
In this way, we can support almost unlimited acknowledgment hole.

Doc: https://docs.google.com/document/d/1uQtyb8t6X04v2vrSrdGWLFkuCkBcGYZbqK8XsVJ4qkU/edit

We do not need to persist the data of all entries every time, only the entries whose Ledger data has changed need to be re-saved. Therefore, we add a wrapper class to record dirty pages.

### Modifications
1. Add a wrapper class
2. Record dirty pages where `add` and `remove`

### Verifying this change
RangeSetWrapperTest.java
1. Make sure dirty pages are recorded correctly
2. Make sure the behavior of wrapper class is the same as other RangeSets

### Documentation

- [x] `no-need-doc` 
(Added configuration items, this part of the document can be automatically generated)

### CODE REVIEW

**Review by codelipenghui — COMMENTED** [review]
LGTM, just left a comment about the config name to keep the same style with configurations we have (*Enabled)

**Review by codelipenghui — APPROVED** [review]

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/RangeSetWrapper.java [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+    private final ManagedLedgerConfig config;
+    private final boolean enableMultiEntry;
+
+    /**
+     * Record which Ledger is dirty.
+     */
+    private final DefaultRangeSet<Long> dirtyLedgers = new LongPairRangeSet.DefaultRangeSet<>(
+            (LongPairConsumer<Long>) (key, value) -> key);
+
+    public RangeSetWrapper(LongPairConsumer<T> rangeConverter, ManagedCursorImpl managedCursor) {
```
[the remark this answers is not in the document]
Comment (315157973) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/RangeSetWrapper.java:50]: Yes, it is useless at this stage. But in the later part, we will use it. In order not to modify the interface, I will pass it in now

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/RangeSetWrapper.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+/**
+ * Wraps other Range classes, and adds LRU, marking dirty data and other features on this basis.
+ * This range set is not thread safety.
+ *
+ * @param <T>
+ */
+public class RangeSetWrapper<T extends Comparable<T>> implements LongPairRangeSet<T> {
+
+    private final LongPairRangeSet<T> rangeSet;
+    private final LongPairConsumer<T> rangeConverter;
```
[the remark this answers is not in the document]
Comment (315157973) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/RangeSetWrapper.java:40]: Yes, it is useless at this stage. But in the later part, we will use it. In order not to modify the interface, I will pass it in now

**Review by hangc0276 — APPROVED** [review]

**Pull request merged**